File: LeetCode-DSA/JulyChallenge_2022/LowestCommonAncestor.cpp
/*Q236. Description: Given a binary tree, find the lowest common ancestor (LCA) of two given nodes in the tree.

According to the definition of LCA on Wikipedia: “The lowest common ancestor is defined between two nodes p and q as the lowest node in T that has both p and q as descendants (where we allow a node to be a descendant of itself).”
 */

class Solution {
public:
    TreeNode* ans = NULL;
    int Traversal(TreeNode* root, TreeNode* p, TreeNode* q)
    {
        if(root == NULL)
            return 0;
        int left = Traversal(root -> left, p, q) ? 1 : 0;
        int right = Traversal(root -> right, p, q) ? 1 : 0;
        int mid = (root == p || root == q) ? 1 : 0;
        // If node is found in left = 1, right = 1, mid = 1 then return root itself
        if(mid + left + right >= 2)
            ans = root;
        // Else, search in left or right subtrees
        return (mid+left+right > 0);
    }
    
    TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) {
        Traversal(root, p, q);
        return ans;
    }
};

File: LeetCode-DSA/LeetCode75/Level1/LowestCommonAncestor.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */

/*
Q235. Description: Given a binary search tree (BST), find the lowest common ancestor (LCA) node of two given nodes in the BST.

According to the definition of LCA on Wikipedia: “The lowest common ancestor is defined between two nodes p and q as the lowest node in T that has both p and q as descendants (where we allow a node to be a descendant of itself).”
*/

class Solution {
    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) 
    {
        // Base case - Empty tree has no nodes
        if(root == null)
            return null;

        if(p.val < root.val && q.val < root.val)
        {
            // Traverse Left side of BST
            return lowestCommonAncestor(root.left, p, q);
        }
        else if(p.val > root.val && q.val > root.val)
            return lowestCommonAncestor(root.right, p, q);
        else
            return root;
    }
}